feat(DOC-2012): document Kafka 4.x client compatibility and KIP-890 (#1614)
redpanda-data/documentation: modules/develop/pages/kafka-clients.adoc

:page-categories: Clients, Development, Kafka Compatibility
:pp: {plus}{plus}
:description: Kafka clients, version 0.11 or later, are compatible with Redpanda. Validations and exceptions are listed.
:page-topic-type: reference
:personas: developer
:learning-objective-1: Identify which Kafka clients are validated with Redpanda
:learning-objective-2: Identify unsupported Kafka features when integrating with Redpanda
// tag::single-source[]

Redpanda is compatible with Apache Kafka® versions 0.11 and later, with specific exceptions noted on this page.

Use this reference to:

* [ ] {learning-objective-1}
* [ ] {learning-objective-2}

== Kafka client compatibility

> Clients developed for Kafka versions 0.11 or later are compatible with Redpanda. The Apache Kafka Java client at version 4.x is validated against Redpanda using the ducktape and chaos test suites; other clients in the table below have not been broadly re-validated at 4.x but remain compatible with the Kafka protocol versions they have always supported. Modern clients auto-negotiate protocol versions or use an earlier protocol version accepted by Redpanda brokers.

Reviewer: > other clients in the table below have not been broadly re-validated at 4.x

I'm not sure I understand what it means. What is "re-validating a client at 4.x"?

> Clients developed for Kafka versions 0.11 or later are compatible with Redpanda. The Apache Kafka Java client at version 4.x is validated against Redpanda using the ducktape and chaos test suites; other clients in the table below have not been broadly re-validated at 4.x but remain compatible with the Kafka protocol versions they have always supported. Modern clients auto-negotiate protocol versions or use an earlier protocol version accepted by Redpanda brokers.

Reviewer 2: ```suggestion
Clients developed for Kafka versions 0.11 or later are compatible with Redpanda. The Apache Kafka Java client at version 4.x is validated against Redpanda using the ducktape and chaos test suites; other clients in the table here remain compatible with the Kafka protocol versions they have always supported. Modern clients auto-negotiate protocol versions or use an earlier protocol version accepted by Redpanda brokers.
```

+
* Quotas per user for bandwidth and API request rates. However, xref:manage:cluster-maintenance/manage-throughput.adoc#client-throughput-limits[quotas per client and per client group] using AlterClientQuotas and DescribeClientQuotas APIs are supported.
endif::[]
+

> * https://cwiki.apache.org/confluence/display/KAFKA/KIP-890[KIP-890^] (Transactions Server-Side Defense): Redpanda does not implement the server-side portion of KIP-890. Kafka 4.x clients use per-transaction epoch bumping to protect against erroneous transaction commits, but fall back to the standard transaction protocol when connecting to Redpanda.

Reviewer: Would it make sense to describe that the Kafka problem that KIP-890 solves is not present in Redpanda due to Raft's atomicity guarantees?

+
* Quotas per user for bandwidth and API request rates. However, xref:manage:cluster-maintenance/manage-throughput.adoc#client-throughput-limits[quotas per client and per client group] using AlterClientQuotas and DescribeClientQuotas APIs are supported.
endif::[]
+

> * https://cwiki.apache.org/confluence/display/KAFKA/KIP-890[KIP-890^] (Transactions Server-Side Defense): Redpanda does not implement the server-side portion of KIP-890. KIP-890 addresses a class of transaction errors specific to Kafka's replication model. Redpanda's Raft-based replication is not susceptible to this issue. Kafka 4.x clients use per-transaction epoch bumping and fall back to the standard transaction protocol when connecting to Redpanda.

Reviewer: > Kafka 4.x clients use per-transaction epoch bumping and fall back to the standard transaction protocol when connecting to Redpanda.

This looks confusing to me. Kafka 4.x clients assume per-transaction epoch bumping only when connected to a server supporting Transactions V2 protocol, which Kafka supports but Redpanda doesn't. Not bumping the epoch on every transaction is part of falling back to the old transactions protocol.

> * https://cwiki.apache.org/confluence/display/KAFKA/KIP-890[KIP-890^] (Transactions Server-Side Defense): Redpanda does not implement the server-side portion of KIP-890, which addresses transaction errors specific to Kafka's replication model. Redpanda's Raft-based replication is not susceptible to this class of errors. When connecting to Redpanda, Kafka 4.x clients detect that Transactions V2 is unsupported and fall back to the original transaction protocol (per-transaction epoch bumping is part of V2 and does not apply).

Reviewer 3: I would remove "raft-based replication" part. That has nothing to do with why Redpanda is not susceptible. A simpler rewording would be .. "Redpanda implementation of transactions is not susceptible.."

> * https://cwiki.apache.org/confluence/display/KAFKA/KIP-890[KIP-890^] (Transactions Server-Side Defense): Redpanda does not implement the server-side portion of KIP-890, which addresses transaction errors specific to Kafka's replication model. Redpanda's Raft-based replication is not susceptible to this class of errors. When connecting to Redpanda, Kafka 4.x clients detect that Transactions V2 is unsupported and fall back to the original transaction protocol (per-transaction epoch bumping is part of V2 and does not apply).

Reviewer 2: ```suggestion
* https://cwiki.apache.org/confluence/display/KAFKA/KIP-890[KIP-890^] (Transactions Server-Side Defense): Redpanda does not implement the server-side portion of KIP-890, which addresses transaction errors specific to Kafka's replication model. Redpanda's implementation of transactions is not susceptible to this class of errors. When connecting to Redpanda, Kafka 4.x clients detect that Transactions V2 is unsupported and fall back to the original transaction protocol (per-transaction epoch bumping is part of V2 and does not apply).
```
